### PR TITLE
Week 4 - Pull 1

### PR DESCRIPTION
**Pull request by jdhang**
This pull request is to review the code done for the first half of week 4.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -17,4 +17,8 @@ def require_user
       redirect_to signin_path
     end
   end
+
+  def set_return_to
```
Comment (brandoncc) [inline app/controllers/application_controller.rb:21]: nice!

File: app/controllers/followings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  def people
+    @followings = Following.where(follower_id: current_user.id)
+  end
+
+  def create
+    @user = User.find(params[:user_id])
+    @following = @user.followings.build(follower_id: current_user.id)
+    if @following.save
+      flash[:notice] = "Following #{@user.full_name}."
+      redirect_to session.delete(:return_to)
```
Comment (brandoncc) [inline app/controllers/followings_controller.rb:14]: Interesting way of handling this, I like it!

File: app/models/user.rb [hunk 3]
```diff
@@ -33,17 +36,15 @@ def queued_video?(video)
...
-
-  def update_queue_item_rating(queue_item, rating)
-    queue_item.video.reviews.where(author: queue_item.user).first.update!(skip_validation: true, rating: rating)
-  end
-
-  def input_rating_empty?(rating)
-    rating == ""
-  end
+    def update_queue_item_rating(queue_item, rating)
+      queue_item.video.reviews.where(author: queue_item.user).first.update!(skip_validation: true, rating: rating)
```
Comment (brandoncc) [inline app/models/user.rb]: Be careful chaining this many methods together. If anything along the way fails, you have an unhandled failure which results in a 500 to your user. Usually you would want to grab the review, then check if it exists, then call update if so. That way you have the ability to handle the failure.

File: app/models/user.rb [hunk 4]
```diff
@@ -33,17 +36,15 @@ def queued_video?(video)
...
+    end
 
+    def input_rating_empty?(rating)
+      rating == ""
```
Comment (brandoncc) [inline app/models/user.rb:51]: You could also do:

``` ruby
rating.blank?
```

Comment (brandoncc) [inline app/models/user.rb:51]: By the way, there is nothing wrong with your way. I just wanted to show you another way :smile:

File: app/views/users/show.html.haml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+              %tr
+                %td
+                  = link_to item.video.title, video_path(item.video)
+                %td
+                  = link_to item.video.category.name, categories_path(item.video.category)
+%section.user_reviews.container
+  .row
+    .col-sm-10.col-sm-offset-1
+      %header
+        %h3= @user.full_name + "'s Reviews (" + @user.reviews.count.to_s + ")"
```
Comment (brandoncc) [inline app/views/users/show.html.haml]: If you use interpolation instead you have the built-in bonus that `@user.full_name` being nil doesn't blow up your app. Currently you would have a `NoMethodError '+' for Nil`.

File: app/views/users/show.html.haml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      %ul
+        - @user.reviews.each do |review|
+          %article.review
+            %li.row
+              .col-sm-2
+                %p
+                  = link_to review.video.title, video_path(review.video)
+                %col= "Rating: " + review.rating.to_s + " / 5"
```
Comment (brandoncc) [inline app/views/users/show.html.haml:37]: Same as above, here is an example:

``` ruby
irb(main):003:0> 'test' + nil + 'test'
TypeError: no implicit conversion of nil into String
    from (irb):3:in `+'
    from (irb):3
    from /Users/brandoncc/.rbenv/versions/2.1.5/bin/irb:11:in `<main>'
irb(main):004:0> "test#{nil}test"
=> "testtest"
```

Comment (brandoncc) [inline app/views/users/show.html.haml:37]: Technically this one is a but different because it is a different error, but the same fix takes care of it.

File: config/database.yml [hunk 7]
```diff
@@ -4,13 +4,13 @@ development:
...
 test:
   adapter: postgresql
   encoding: unicode
   database: myflix_test
   pool: 5
-  username: jdhang
-  password: test
\ No newline at end of file
+  username: jhang
+  password:
```
Comment (brandoncc) [inline config/database.yml]: Your `database.yml` file should not be committed to your repo. You should commit a `database.yml.sample` which contains an outline of what is expected, but never your live yml file.

File: db/migrate/20150203165922_create_followings.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class CreateFollowings < ActiveRecord::Migration
+  def change
+    create_table :followings do |t|
+      t.timestamps null:false
```
Comment (brandoncc) [inline db/migrate/20150203165922_create_followings.rb:4]: `null: false` is not necessary because Rails creates the data with the timestamps. It will not make it into your db without timestamps.

File: db/migrate/20150203165922_create_followings.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+      t.integer :user_id, :follower_id, null:false
```
Comment (brandoncc) [inline db/migrate/20150203165922_create_followings.rb:5]: Be careful putting the requirements at the db level. While this provides an extra layer of protection, it also introduces another level you need to remember to change if you decide to change how your app works.

File: spec/controllers/followings_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
+require 'spec_helper'
+
+describe FollowingsController do
+  before(:each) do
+    request.env["HTTP_REFERER"] = "previous_page"
+  end
+  describe "GET people" do
+    context "with authenticated users" do
+      let(:following) { Fabricate(:following, follower_id: current_user.id) }
```
Comment (brandoncc) [inline spec/controllers/followings_controller_spec.rb:6]: When you only have a single spec, using `let` and `before` seems like more work than just putting the code in the spec. While `let` and `before` make your code nicer to look at, they also introduce abstraction that can make your code harder to follow. I speak from experience, I am currently dealing with overuse of both in my own app.

File: spec/controllers/sessions_controller_spec.rb [hunk 11]
```diff
@@ -1,105 +1,73 @@
...
         session[:user_id] = Fabricate(:user).id
         get :destroy
       end
-    
       it "resets session[:user_id]" do
         expect(session[:user_id]).to be_nil
       end
-
       it "sets flash success" do
         expect(flash[:success]).to_not be_blank
```
Comment (brandoncc) [inline spec/controllers/sessions_controller_spec.rb]: You can also use the positive expectation `be_present`.

File: spec/fabricators/following_fabricator.rb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Fabricator(:following) do
```
Comment (brandoncc) [inline spec/fabricators/following_fabricator.rb:1]: You can also do

``` ruby
Fabricator(:following) do
  user
end
```

This will give you the same result. That is shorthand for this:

``` ruby
Fabricator(:following) do
  user { Fabricate(:user) }
end
```

File: app/models/user.rb (hunk 3 above)
Comment (jdhang) [inline app/models/user.rb]: That's a good point, I will separate finding the review and if it exists then updating it.  Thanks!

File: app/views/users/show.html.haml (hunk 5 above)
Comment (jdhang) [inline app/views/users/show.html.haml]: That's a very good point, will modify to prevent this error from happening! Thanks!

File: config/database.yml (hunk 7 above)
Comment (jdhang) [inline config/database.yml]: Ah yes, I did question this when I was committing, it felt wrong, probably should have gone with my gut. Will remove and add a sample file. Thanks!

File: spec/controllers/followings_controller_spec.rb (hunk 10 above)
Comment (jdhang) [inline spec/controllers/followings_controller_spec.rb:6]: So the `before` was to satisfy the `sessions[:return_to]` feature of the app. It would give me an error since there was not previous page to reference. Do you think there is a better way to implement this or just add the `before` for each `describe` spec? Thanks!

File: spec/controllers/sessions_controller_spec.rb (hunk 11 above)
Comment (jdhang) [inline spec/controllers/sessions_controller_spec.rb]: Oh thanks! Yeah I like positive expectations better so I will change this.

File: spec/fabricators/following_fabricator.rb (hunk 12 above)
Comment (jdhang) [inline spec/fabricators/following_fabricator.rb:1]: Great! I didn't know that, saves me some typing, thanks!

File: spec/controllers/followings_controller_spec.rb (hunk 10 above)
Comment (brandoncc) [inline spec/controllers/followings_controller_spec.rb:6]: `before` blocks are run right before the `it` block. That means you can combine the two yourself as well.  For example, you could rewrite this context like this:

``` ruby
context "with authenticated users" do
  it "sets @followings" do
    request.env["HTTP_REFERER"] = "previous_page"
    following = Fabricate(:following, follower_id: current_user.id)
    set_current_user
    get :people
    expect(assigns(:followings)).to eq([following])
  end
end
```

Here is an example of why before can be a bit hard to follow when you are reading code. I missed that first line (setting the referrer) until I re-read the question before posting this reply. I had all of the other lines, but didn't even see that one.

File: spec/fabricators/following_fabricator.rb (hunk 12 above)
Comment (brandoncc) [inline spec/fabricators/following_fabricator.rb:1]: You can scan the docs for the gem in about 5 minutes, if you are so inclined. A lot of good information here: http://www.fabricationgem.org/

The "Defining Fabricators" tab applies to this specifically.

File: spec/controllers/followings_controller_spec.rb (hunk 10 above)
Comment (jdhang) [inline spec/controllers/followings_controller_spec.rb:6]: Oh okay, that makes more sense, so I will modify it to be more explicit

File: spec/fabricators/following_fabricator.rb (hunk 12 above)
Comment (jdhang) [inline spec/fabricators/following_fabricator.rb:1]: Will do!
